mlflow/llama_index/pyfunc_wrapper.py

```diff
@@ -0,0 +1,159 @@
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from llama_index.core import QueryBundle
+from llama_index.core.llms import ChatMessage
+
+from mlflow.models.utils import _convert_llm_input_data
+
+CHAT_ENGINE_NAME = "chat"
+QUERY_ENGINE_NAME = "query"
+RETRIEVER_ENGINE_NAME = "retriever"
+SUPPORTED_ENGINES = {CHAT_ENGINE_NAME, QUERY_ENGINE_NAME, RETRIEVER_ENGINE_NAME}
+
+_CHAT_MESSAGE_PARAMETER_NAMES = {"query", "message"}
+_CHAT_MESSAGE_HISTORY_PARAMETER_NAMES = {"messages", "chat_history"}
+
+
+class _LlamaIndexModelWrapperBase:
+    def __init__(
+        self,
+        index,
+        model_config: Optional[Dict[str, Any]] = None,
+    ):
+        self.index = index
+        self.model_config = model_config or {}
+        self.predict_callable = self._build_engine_method()
+
+    def _build_engine_method(self) -> Callable:
+        raise NotImplementedError
+
+    def _format_predict_input(self, data):
+        raise NotImplementedError
+
+    def _do_inference(self, input, params: Optional[Dict[str, Any]]) -> Dict:
+        """
+        Perform engine inference on a single engine input e.g. not an iterable of
+        engine inputs. The engine inputs must already be preprocessed/cleaned.
+        """
+
+        if isinstance(input, dict):
+            return self.predict_callable(**input, **(params or {}))
+        else:
+            return self.predict_callable(input, **(params or {}))
+
+    def predict(self, data, params: Optional[Dict[str, Any]] = None) -> Union[List[str], str]:
+        data = self._format_predict_input(data)
+
+        if isinstance(data, list):
+            return [self._do_inference(x, params) for x in data]
+        else:
+            return self._do_inference(data, params)
+
+
+class ChatEngineWrapper(_LlamaIndexModelWrapperBase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.engine_type = CHAT_ENGINE_NAME
+
+    def _build_engine_method(self) -> Callable:
+        return self.index.as_chat_engine(**self.model_config).chat
+
+    @staticmethod
+    def _parse_dict_as_chat_message_objects(data: dict) -> dict:
+        message_matches = (name for name in _CHAT_MESSAGE_PARAMETER_NAMES if name in data)
+        if (key := next(message_matches, None)) and data.get(key):
```

Comment:
This part is a bit hard to follow, can we simply construct a for loop?
```
for key in CHAT_MESSAGE_PARAMETER_NAMES:
    if v := data.get(key) and not isinstance(v, str):
```

Reply:
No idea why this is written like this 🙃 

Thanks